pkg/controller/direct/compute/refs.go

```diff
@@ -0,0 +1,475 @@
+		}
+		return ref, nil
+	}
+
+	if ref.Name == "" {
+		return nil, fmt.Errorf("must specify either name or external on reference")
+	}
+
+	key := types.NamespacedName{
+		Namespace: ref.Namespace,
+		Name:      ref.Name,
+	}
+	if key.Namespace == "" {
+		key.Namespace = src.GetNamespace()
+	}
+
+	computeNetwork, err := resolveResourceName(ctx, reader, key, schema.GroupVersionKind{
+		Group:   "compute.cnrm.cloud.google.com",
+		Version: "v1beta1",
+		Kind:    "ComputeNetwork",
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	// targetField: self_link
+	// See compute servicemappings for details
+	return &refs.ComputeNetworkRef{
+		External: computeNetwork.GetSelfLink()}, nil
+}
+
+func ResolveComputeSubnetwork(ctx context.Context, reader client.Reader, src client.Object, ref *refs.ComputeSubnetworkRef) (*refs.ComputeSubnetworkRef, error) {
+	if ref == nil {
+		return nil, nil
+	}
+
+	if ref.External != "" {
+		if ref.Name != "" {
+			return nil, fmt.Errorf("cannot specify both name and external on reference")
+		}
+		return ref, nil
+	}
+
+	if ref.Name == "" {
+		return nil, fmt.Errorf("must specify either name or external on reference")
+	}
+
+	key := types.NamespacedName{
+		Namespace: ref.Namespace,
+		Name:      ref.Name,
+	}
+	if key.Namespace == "" {
+		key.Namespace = src.GetNamespace()
+	}
+
+	computeSubnetwork, err := resolveResourceName(ctx, reader, key, schema.GroupVersionKind{
+		Group:   "compute.cnrm.cloud.google.com",
+		Version: "v1beta1",
+		Kind:    "ComputeSubnetwork",
+	})
+	if err != nil {
+		return nil, err
+	}
+	// targetField: self_link
+	// See compute servicemappings for details
+	return &refs.ComputeSubnetworkRef{
+		External: computeSubnetwork.GetSelfLink(),
+	}, nil
+}
+
+func ResolveComputeAddress(ctx context.Context, reader client.Reader, src client.Object, ref *refs.ComputeAddressRef) (*refs.ComputeAddressRef, error) {
+	if ref == nil {
+		return nil, nil
+	}
+
+	if ref.External != "" {
+		if ref.Name != "" {
+			return nil, fmt.Errorf("cannot specify both name and external on reference")
+		}
+		return ref, nil
+	}
+
+	if ref.Name == "" {
+		return nil, fmt.Errorf("must specify either name or external on reference")
+	}
+
+	key := types.NamespacedName{
+		Namespace: ref.Namespace,
+		Name:      ref.Name,
+	}
+	if key.Namespace == "" {
+		key.Namespace = src.GetNamespace()
+	}
+
+	computeAddress, err := resolveResourceName(ctx, reader, key, schema.GroupVersionKind{
+		Group:   "compute.cnrm.cloud.google.com",
+		Version: "v1beta1",
+		Kind:    "ComputeAddress",
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// targetField: address
+	// See compute servicemappings for details
+	address, _, _ := unstructured.NestedString(computeAddress.Object, "spec", "address")
+	if address == "" {
+		return nil, fmt.Errorf("cannot get address for referenced %s %v (spec.address is empty)", computeAddress.GetKind(), computeAddress.GetNamespace())
+	}
+	return &refs.ComputeAddressRef{
+		External: address}, nil
```

Comment:
Will be interesting to see if this is the "primary" external ref format .... TBD!